### DISCUSSION TITLE
DeepL API, Crowdin, Markdown Translator and etc.

### DISCUSSION THREAD

**Discussion by rsipakov** [timlrx/tailwind-nextjs-starter-blog#179] [Ideas]
As it was previously proposed the idea to integrate automatic translation (branch demo/translate), it is proposed to discuss:

- [DeepL API](https://www.deepl.com/docs-api/),
- [Markdown Translator](https://www.npmjs.com/package/markdown-translator)
- [Using Crowdin on example with Docusaurus](https://docusaurus.io/docs/i18n/crowdin) 

in this thread.

**Comment by GautierArcin** [comment]
To add to the Idea, it seems kind of very difficult.

There are two different things to think about:

1. The translation API and its capabilities.

2. How we do the translation.

For 1., I didn't found any API that are able  to convert `.mdx` files. 
I did found some possibilities for `.md` files, but in the end, it won't probably be deepl. Deepl tends to do a lot of mistakes with markdown translation.
Unfortunately, I did not find any satisfying way of doing it for free. The best way  to do it for free would be to create an intermediary library that would parse only text (without formatting) in the correct format (not too many characters for the free version) and making sure to correctly manage error codes.
For paid versions, there is [this package](https://www.npmjs.com/package/markdown-translator) that works with Azure.

For 2., we don't need to do again the translation of each article at build time.
The best way of doing so would be to store the `md5` of each markdown article in a central `translate.json` file. 
When we generate a build, we check the `.md5` of each `.md(x)` files in the `defaultLocale`. If it changes, or is absent, we do the translation. Otherwise, we don't.
I'm still trying to think of a way of implementing that behavior with Vercel (Vercel won't be able to cache files between build, which nullify totally what I just wrote...). I'm open to any ideas on that part.

In the meantime, I'm defaulting to manually translate each file in my blog. 😅

**Reply by rsipakov** [reply]
I found one more idea using [Crowdin](https://docusaurus.io/docs/i18n/crowdin)
